File: src/DefaultAzureCredential.PythonClient/Authentication.py
from azure.core.exceptions import ResourceNotFoundError
from azure.storage.blob import BlobServiceClient
from azure.identity import DefaultAzureCredential
import jwt


class Authentication:
    async def get_upn_from_token(self, default_azure_credential: DefaultAzureCredential, scope: str) -> None:
        try:
            access_token = await default_azure_credential.get_token(scope)
            jwt_security_token_handler = jwt.JWT()
            jwt_security_token = jwt_security_token_handler.decode(access_token.token)
            print(f"UPN from JWT Token: {jwt_security_token.get('upn')}")
        except Exception as ex:
            print(f"Exception: {ex}")

    async def list_blob_storage_container(self, default_azure_credential: DefaultAzureCredential, storage_account_name: str) -> None:
        blob_uri = f"https://{storage_account_name}.blob.core.windows.net"
        blob_service_client = BlobServiceClient(blob_uri, credential=default_azure_credential)
        try:
            containers = blob_service_client.list_containers()
            async for container in containers:
                print(f"...Container Name: {container.name}")
            print("Storage access succeeded")
        except ResourceNotFoundError as ex:
            print(f"Resource not found error: {ex}")
        except Exception as ex:
            print(f"Exception: {ex}")

File: src/DefaultAzureCredential.Client.CSharp/Authentication.cs
using System.IdentityModel.Tokens.Jwt;
using Azure; 
using Azure.Core;
using Azure.Identity;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;
using Microsoft.Azure.Devices;
using Microsoft.IdentityModel.Tokens;

internal class Authentication
{
    async internal Task GetUpnFromToken(DefaultAzureCredential defaultAzureCredential, string scope)
    {
        try {
            AccessToken accessToken = await defaultAzureCredential.GetTokenAsync(
                new TokenRequestContext(new string[] {scope})
            );
            JwtSecurityTokenHandler jwtSecurityTokenHandler = new JwtSecurityTokenHandler(); 
            JwtSecurityToken jwtSecurityToken = (JwtSecurityToken)jwtSecurityTokenHandler.ReadToken(accessToken.Token);
            Console.Write($"UPN from JWT Token: {jwtSecurityToken.Claims.First(c => c.Type=="upn").Value}"); 
        } catch (Exception exE) {
            Console.WriteLine($"Exception: {exE.Message}");
        }
    }


    internal async Task ListBlobStorageContainer(DefaultAzureCredential defaultAzureCredential, string storageAccountName)
    {

        Uri blobUri = new Uri($"https://{storageAccountName}.blob.core.windows.net");

        BlobServiceClient blobServiceClient = new BlobServiceClient(blobUri, defaultAzureCredential);   

        try {
            IAsyncEnumerable<Page<BlobContainerItem>> resultPages = 
                blobServiceClient.GetBlobContainersAsync()
                .AsPages(default, 10);

            await foreach (Page<BlobContainerItem> containerPage in resultPages)
            {
                foreach (BlobContainerItem containerItem in containerPage.Values)
                {
                    Console.WriteLine($"...Container Name: {containerItem.Name}"); 
                }
            }
            Console.WriteLine("Storage access succeeded");
        } catch (AuthenticationFailedException exE){
            Console.WriteLine($"Authentication failed: {exE.Message}");
        }
    }
}